pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,673 @@
+
+	jobListerSynced     cache.InformerSynced
+	cronJobListerSynced cache.InformerSynced
+
+	// now is a function that returns current time, done to facilitate unit tests
+	now func() time.Time
+}
+
+// NewControllerV2 creates and initializes a new Controller.
+func NewControllerV2(jobInformer batchv1informers.JobInformer, cronJobsInformer batchv1beta1informers.CronJobInformer, kubeClient clientset.Interface) (*ControllerV2, error) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		if err := ratelimiter.RegisterMetricAndTrackRateLimiterUsage("cronjob_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter()); err != nil {
+			return nil, err
+		}
+	}
+
+	jm := &ControllerV2{
+		queue:    workqueue.NewNamedDelayingQueue("cronjob"),
+		recorder: eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cronjob-controller"}),
+
+		jobControl:     realJobControl{KubeClient: kubeClient},
+		cronJobControl: &realCJControl{KubeClient: kubeClient},
+
+		jobLister:     jobInformer.Lister(),
+		cronJobLister: cronJobsInformer.Lister(),
+
+		jobListerSynced:     jobInformer.Informer().HasSynced,
+		cronJobListerSynced: cronJobsInformer.Informer().HasSynced,
+		now:                 time.Now,
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    jm.addJob,
+		UpdateFunc: jm.updateJob,
+		DeleteFunc: jm.deleteJob,
+	})
+
+	cronJobsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+		UpdateFunc: jm.updateCronJob,
+		DeleteFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+	})
+
+	return jm, nil
+}
+
+// Run starts the main goroutine responsible for watching and syncing jobs.
+func (jm *ControllerV2) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer jm.queue.ShutDown()
+
+	klog.Infof("Starting cronjob controller v2")
+	defer klog.Infof("Shutting down cronjob controller v2")
+
+	go jm.enqueueAllCronJobsWithRetries(retriesForEnqueingAllCronJobs)
+
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, jm.jobListerSynced, jm.cronJobListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(jm.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// enqueueAllCronJobsWithRetries enqueues all the cronjobs and is thread-safe.
+func (jm *ControllerV2) enqueueAllCronJobsWithRetries(retries int) {
+	cronjobs := jm.listCronJobsWithRetries(retries)
+	for _, cronjob := range cronjobs {
+		jm.enqueueController(cronjob)
+	}
+	return
+}
+
+// listCronJobsWithRetries tries to list the cronjobs with retries upon hitting an error
+func (jm *ControllerV2) listCronJobsWithRetries(retries int) []*batchv1beta1.CronJob {
```

Comment:
Since this is a one time go routine, I was not sure what would happen if we hit a transient api-server. The thought was that we would want to retry a number of tries and hope that the transient apiserver error goes away, if not we fail at enqueueing the crons at the start of the controller

Reply:
This is not an elegant solution, however most solutions that come to my mind are little complex and would take them as follow on PRs. If anything simple comes to mind, open to adding it now

Reply:
You already are waiting for caches sync (https://github.com/kubernetes/kubernetes/pull/93370/files#diff-3ac32b4c4bbfd547f2224dda02e1b422629791f723459a6cfb6a1a0670519975R131), and here you're only reading from that so no need for retries.

Reply:
Maceij pointed out that we do not need the retries, because we are waiting for caches to be synced anyway. I missed that safety net. +1 on removing the retry logic

Reply:
Done